icu4c/source/common/uloc.cpp

```diff
@@ -1324,55 +1324,75 @@ _getVariant(const char* localeID,
+            UErrorCode& status) {
+    if (U_FAILURE(status)) return;
     if (pEnd != nullptr) { *pEnd = localeID; }
 
+    // Reasonable upper limit for variants
+    // There are no strict limitation of the syntax of variant in the legacy
+    // locale format. If the locale is constructed from unicode_locale_id
+    // as defined in UTS35, then we know each unicode_variant_subtag
+    // could have max length of 8 ((alphanum{5,8} | digit alphanum{3})
+    // 179 would allow 20 unicode_variant_subtag with sep in the
+    // unicode_locale_id
+    // 8*20 + 1*(20-1) = 179
+    constexpr int32_t MAX_VARIANTS_LENGTH = 179;
+
     /* get one or more variant tags and separate them with '_' */
-    if(_isIDSeparator(prev)) {
+    int32_t index = 0;
+    if (_isIDSeparator(prev)) {
         /* get a variant string after a '-' or '_' */
-        while(!_isTerminator(*localeID)) {
+        for (index=0; !_isTerminator(localeID[index]); index++) {
             if (needSeparator) {
                 if (sink != nullptr) {
                     sink->Append("_", 1);
                 }
                 needSeparator = false;
             }
             if (sink != nullptr) {
-                char c = (char)uprv_toupper(*localeID);
+                char c = (char)uprv_toupper(localeID[index]);
                 if (c == '-') c = '_';
                 sink->Append(&c, 1);
             }
-            hasVariant = true;
-            localeID++;
+            if (index > MAX_VARIANTS_LENGTH) {
+                status = U_ILLEGAL_ARGUMENT_ERROR;
+                return;
+            }
         }
-        if (pEnd != nullptr) { *pEnd = localeID; }
+        if (pEnd != nullptr) { *pEnd = localeID+index; }
     }
 
     /* if there is no variant tag after a '-' or '_' then look for '@' */
-    if(!hasVariant) {
-        if(prev=='@') {
+    if (index == 0) {
+        if (prev=='@') {
             /* keep localeID */
         } else if((localeID=locale_getKeywordsStart(localeID))!=nullptr) {
             ++localeID; /* point after the '@' */
         } else {
             return;
         }
-        while(!_isTerminator(*localeID)) {
+        for(; !_isTerminator(localeID[index]); index++) {
             if (needSeparator) {
                 if (sink != nullptr) {
                     sink->Append("_", 1);
                 }
                 needSeparator = false;
             }
             if (sink != nullptr) {
-                char c = (char)uprv_toupper(*localeID);
+                char c = (char)uprv_toupper(localeID[index]);
                 if (c == '-' || c == ',') c = '_';
                 sink->Append(&c, 1);
             }
-            localeID++;
+            if (index > MAX_VARIANTS_LENGTH) {
+                status = U_ILLEGAL_ARGUMENT_ERROR;
+                return;
+            }
         }
-        if (pEnd != nullptr) { *pEnd = localeID; }
+        if (pEnd != nullptr) { *pEnd = localeID + index; }
+    }
+    if (index > MAX_VARIANTS_LENGTH) {
```

Comment:
It seems to me that this could never happen. Are there any circumstances under which this code could actually be reached? If not, it won't ever be used and can be deleted.

Reply:
yes, you are right. It is hard for me to see this is will not happen with a small screen but after carefully looking up/down that is surely the case and I now remove this.

Reply:
This happen when  _isTerminator(localeID[MAX_VARIANTS_LENGTH+1]) is true.
Inside the loop, in the previous iteration, index == MAX_VARIANTS_LENGTH so it does not error return by the "if (index > MAX_VARIANTS_LENGTH)" check, then the loop go to the next iteration by index++ which make index is MAX_VARIANTS_LENGTH +1 first, then it call _isTerminator(localeID[index]) , but that return true cause the loop terminate. at this point index is MAX_VARIANTS_LENGTH +1 which is > MAX_VARIANTS_LENGTH.